xmtp_mls/README.md

```diff
@@ -122,41 +116,58 @@ The [following diagram](https://app.excalidraw.com/s/4nwb0c8ork7/6pPH1kQDoj3) il
 1. If no conflicts: Publish welcome messages to new members.
-   If conflicts: Go back to step 4 and try again
+   If conflicts: Go back to step 2 and try again (reset the intent's state to `TO_SEND` and clear the `data` and `post_commit_data` fields)
 
 ### Remove members from a group
 
 Simplified high level flow for removing members from a group:
 
 1. Create a `group_intent` for removing the members
-1. Sync the state of the group with the network
-1. Convert the intent into concrete commit for the current epoch
+1. Convert the intent (and any pending proposals) into concrete commit for the current epoch
+   1. If there are welcome messages, write them to the `post_commit_data` field for later
 1. Publish commit to the network
 1. Sync the state of the group with the network
-1. If no conflicts: Done.
-   If conflicts: Go back to step 3 and try again
+1. If no conflicts: Done. If there are welcome messages in the `post_commit_data` field, publish them.
+   If conflicts: Go back to step 2 and try again (reset the intent's state to `TO_SEND` and clear the `data` and `post_commit_data` fields)
 
 ### Send a message
 
 Simplified high level flow for sending a group message:
 
 1. Create a `group_intent` for sending the message
+1. Commit any pending proposals
+   1. If there are welcome messages, write them to the `post_commit_data` field for later
 1. Convert the intent into a concrete message for the current epoch
 1. Publish message to the network
 1. Sync the state of the group with the network (can be debounced or otherwise only done periodically)
-1. If no conflicts: Mark the message as committed. If conflicts: Go back to step 2.
+1. If no conflicts: Mark the message as committed. If there are welcome messages in the `post_commit_data` field, publish them.
+   If conflicts: Go back to step 2 and try again (reset the intent's state to `TO_SEND` and clear the `data` and `post_commit_data` fields)
+
+### Syncing group state
+
+The server maintains an inbound topic for each group, and a single inbound topic for the client's identity. For each topic, the client maintains a row that stores `last_synced_payload_id` and `lock_sync_until_ns` fields.
+
+1. In a single transaction, validate that `lock_until_ns` is not set to a value greater than `now()`, set it to a timeframe in the future, and fetch the `last_message_timestamp_ns`
+1. Fetch all payloads greater than the timestamp from the server
+1. Sequentially process each payload. For each payload, update the `last_message_timestamp_ns` and any corresponding database writes for that payload in a single transaction
+1. When the sync is complete, release the lock by setting `lock_until_ns` to 0
+
+This flow will be similar regardless of if the sync happens via a poll-based or subscription-based mechanism. For a subscription-based mechanism, the lock will be obtained at the start of the subscription, and extended on a heartbeat time interval, until the subscription is closed.
```

Comment:
In the absence of really robust subscriptions (which I think is going to be quite hard to implement for V1) my recommendation would be to not adjust any locks on the subscription and allow state syncs to happen in parallel.

Reply:
If subscriptions are not robust, I feel like it causes a lot of downstream problems.
1) If a commit message is lost, subsequent payloads are unreadable. You can trigger a pull at this point, but it's added complexity.
2) If we don't lock syncs while subscriptions are in progress, then we have concurrent processing of the same payloads, meaning one side or the other will fail. It becomes unclear if failures are due to bad payloads/missing epoch messages, or if they are due to parallel processing.
3) It's not the best dev experience for consumers of the SDK.

I have some suggestions:
1) We could make subscriptions more robust. We could pass the `last_synced_payload_timestamp` to the server when initiating subscriptions. Alternatively, we could have the server record the previous payload timestamp on each payload, so that the client can detect when it is missing payloads and re-sync.
2) We could omit subscriptions from the initial MLS milestones and fix them up in later milestones.

Reply:
I do like the third option. I hadn't considered adding the `previous_payload_timestamp` to all our streamed envelopes.

Given that all options look like option 1 in some cases, I wonder if we start with that to unblock everything and give us streaming methods quite easily. Then we can layer on option 2 or 3 depending on how we feel about the scope of changes.

Reply:
That's a really good framing, thanks for writing it up. Okay, this makes sense to me!